modules/renter/contractor/contracts.go

```diff
@@ -0,0 +1,445 @@
+			newCost = prevCost.Mul64(2)
+		} else {
+			prevBase := contract.TotalCost.Sub(contract.ContractFee).Sub(contract.TxnFee).Sub(contract.SiafundFee)
+			// The amount of money that the contract started with should not be
+			// less than the amount of money remaining in the contract, but
+			// double check just to be sure.
+			if prevBase.Cmp(contract.RenterFunds()) < 0 {
+				build.Critical("A contracts base funds is smaller than it's available funds:", prevBase, contract.RenterFunds())
+			}
+			prevCost := prevBase.Sub(contract.RenterFunds())
+			newCost = prevCost.Mul64(3).Div64(2) // Set the renterFunds to 50% more than the amount of money spent in the last cylce.
+		}
+
+		// Verify that the new cost is at least enough to cover all the existing
+		// data and some extra.
+		timeExtension := uint64(contractsEndHeight - contract.LastRevision.NewWindowEnd) // TODO: May need to add host.WindowSize too.
+		// Determine how much is going to be spend immediately on existing
+		// storage.
+		baseCost := host.StoragePrice.Mul64(contract.LastRevision.NewFileSize).Mul64(timeExtension)
+		// Make sure that the renterFunds are at least 50% greater than the base
+		// cost.
+		minFunds := baseCost.Mul64(3).Div64(2)
+		if minFunds.Cmp(newCost) < 0 {
+			newCost = minFunds
+		}
+		// Make sure there is at least a minimum number of coins in the
+		// contract.
+		if newCost.Cmp(minContractFunds) < 0 {
+			newCost = minContractFunds
+		}
+
+		// Calculate the desired collateral.
+		collateral := c.collateralFromCost(newCost, host)
+
+		c.managedLockContract(contract.ID)
+		c.managedRenewContract(contract, host, newCost, collateral, contractsEndHeight)
+		c.managedUnlockContract(contract.ID)
+
+		// Soft sleep between contract formation.
+		select {
+		case <-c.tg.StopChan():
+			return
+		case <-time.After(contractFormationInterval):
+		}
+	}
+
+	// Determine how many new contracts we want to form. Also grab the list of
+	// hosts that we already have, so that we do not form any duplicate
+	// contracts.
+	c.mu.Lock()
+	existingHosts := make([]types.SiaPublicKey, 0, len(c.contracts))
+	wantedHosts := c.allowance.Hosts
+	for _, contract := range c.contracts {
+		existingHosts = append(existingHosts, contract.HostPublicKey)
+	}
+	c.mu.Unlock()
+
+	// Select a bunch of new hosts from the database.
+	selectionSize := int(wantedHosts*3 + 10)                 // Fine to have an abundance of hosts.
+	hosts := c.hdb.RandomHosts(selectionSize, existingHosts) // Fine to have an abundance of hosts.
+
+	// Filter any hosts that do not have enough storage remaining.
+	i := 0
+	for {
+		if i >= len(hosts) {
+			break
+		}
+		if hosts[i].RemainingStorage < storageRemainingThreshold {
+			hosts = append(hosts[:i], hosts[i+1:]...)
+		} else {
+			i++
+		}
+
+	}
+
+	// Form contracts with the hosts until we have the desired number of total
+	// contracts.
+	i = 0
+	for numGoodContracts < wantedHosts && i < len(hosts) {
```

Comment:
I would probably do `for _, host := range hosts` and then add the `numGoodContracts < wantedHosts` check inside the loop. That also lets you skip the last sleep.